katcbfsim/transport.py

```diff
@@ -44,53 +51,65 @@ class SpeadTransport(object):
-            e = endpoints[i * len(endpoints) // stream.n_substreams]
-            if ibv:
-                sender = spead2.send.asyncio.UdpIbvStream(
-                    spead2.ThreadPool(),
-                    config,
-                    spead2.send.UdpIbvConfig(
-                        endpoints=[(e.host, e.port)],
-                        interface_address=ifaddr,
-                        ttl=4           # For MeerKAT layer 3 switching
-                    )
+        out_rate = in_rate * 1.1 / stream.n_substreams * my_n_substreams
+        config = spead2.send.StreamConfig(
+            max_heaps=my_n_substreams,
+            rate=out_rate,
+            max_packet_size=max_packet_size)
+        my_endpoints = [
+            endpoints[i * len(endpoints) // stream.n_substreams]
+            for i in range(first_substream, last_substream)
+        ]
+        # spead2 takes a tuple, not an Endpoint object
+        spead2_endpoints = [
+            (endpoint.host, endpoint.port)
+            for endpoint in my_endpoints
+        ]
+        if ibv:
+            sender = spead2.send.asyncio.UdpIbvStream(
+                spead2.ThreadPool(),
+                config,
+                spead2.send.UdpIbvConfig(
+                    endpoints=spead2_endpoints,
+                    interface_address=ifaddr,
+                    ttl=4           # For MeerKAT layer 3 switching
                 )
-            else:
-                kwargs = {}
-                if ifaddr is not None:
-                    kwargs['interface_address'] = ifaddr
-                    kwargs['ttl'] = 4   # For MeerKAT layer 3 switching
-                sender = spead2.send.asyncio.UdpStream(
-                    spead2.ThreadPool(), [(e.host, e.port)], config, **kwargs)
-            channel0 = i * stream.n_channels // stream.n_substreams
-            channel1 = (i + 1) * stream.n_channels // stream.n_substreams
-            self._substreams.append(Substream(
-                sender=sender, endpoint=e, channel_range=slice(channel0, channel1)))
-            self._substreams[-1].sender.set_cnt_sequence(i, stream.n_substreams)
+            )
+        else:
+            kwargs = {}
+            if ifaddr is not None:
+                kwargs['interface_address'] = ifaddr
+                kwargs['ttl'] = 4   # For MeerKAT layer 3 switching
+            sender = spead2.send.asyncio.UdpStream(
+                spead2.ThreadPool(), spead2_endpoints, config, **kwargs)
+        sender.set_cnt_sequence(server_id, n_servers)
+        self._substreams = [
+            Substream(
+                sender=sender,
+                endpoint=endpoint,
+                channel_range=slice(
+                    i * stream.n_channels // stream.n_substreams,
+                    (i + 1) * stream.n_channels // stream.n_substreams
+                ),
+                index=i - first_substream
+            )
+            for i, endpoint in zip(range(first_substream, last_substream), my_endpoints)
```

Comment:
Is enumerate() frowned upon these days ? len(my_endpoints) == last_substream-first_substream anyway, so the zip appears unneeded.

Reply:
I'd forgotten that enumerate took an optional start. I'll change to use that.